feature/task 1

zaurbekzet opened the pull request.

zaurbekzet committed:
Add Location struct

zaurbekzet committed:
Implement LocationService

zaurbekzet committed:
Add json tags

zaurbekzet committed:
Add some initial data

zaurbekzet committed:
Implement LocationCommander

zaurbekzet committed:
Add logistic/location support

zaurbekzet committed:
Add IDE config folders to .gitignore

popovid commented on internal/app/commands/logistic/commander.go:
```diff
@@ -0,0 +1,47 @@
+	locationCmd "github.com/ozonmp/omp-bot/internal/app/commands/logistic/location"
+	"github.com/ozonmp/omp-bot/internal/app/path"
+	locationSrv "github.com/ozonmp/omp-bot/internal/service/logistic/location"
+	"log"
```
предложу следовать гошному стилю в оформлении импортов https://github.com/golang/go/wiki/CodeReviewComments#imports

zaurbekzet replied:
Настроил автоматический запуск goimports при сохранении файлов в GoLand. Исправил: d8ac47d7d318839faf7fa07041a955e8c67d7d7b

popovid commented on internal/app/commands/logistic/location/command_help.go:
```diff
@@ -0,0 +1,16 @@
+
+func (c *LogisticLocationCommander) Help(inputMsg *tgbotapi.Message) {
+	msgText := `/help__logistic__location — print list of commands
+/get__logistic__location — get location
```
может добавим сразу в help, что нужно добавлять id для подобных запросов?

zaurbekzet replied:
Добавил форматы всех команд: 7416b0ed0bf40ad44cceeb0942566495bae10c63

popovid commented on internal/app/commands/logistic/location/command_get.go:
```diff
@@ -0,0 +1,31 @@
+func (c *LogisticLocationCommander) Get(inputMsg *tgbotapi.Message) {
+	args := inputMsg.CommandArguments()
+
+	id, err := strconv.Atoi(args)
```
может сразу заменить на strconv.ParseUint, раз потом uint и требуется ?

zaurbekzet replied:
Заменил: 3df50ad6c43b661e4b461f7eac4755ac771fb221

popovid commented on internal/app/commands/logistic/location/command_list.go:
```diff
@@ -0,0 +1,59 @@
+	locations, err := c.locationService.List(cursor, limit)
+	if err != nil {
+		log.Printf("%s.List: failed to list locations: %v", logPrefix, err)
+		//c.sendMessage(inputMsg.Chat.ID, "Error occurred", logPrefix+".List")
```
наверное стоит удалить, или раскомментить, если все-таки нужно)

zaurbekzet replied:
Раскомментировал и исправил обработку ситуаций с пустыми списками и отображением кнопки _Next_ в случае, когда больше нечего показывать: a740fedbfefbae0ba07b89f5b47c6bf59aace93f

popovid commented on internal/model/logistic/location.go:
```diff
@@ -0,0 +1,14 @@
+	ID        uint64  `json:"id,omitempty"`
+	Latitude  float64 `json:"latitude"`
+	Longitude float64 `json:"longitude"`
+	Title     string  `json:"title"`
```
в целом, если в json у полей такое же название, можно и не ставить теги
Это не ошибка, просто FYI)

zaurbekzet replied:
Тэги добавил, чтобы имена ключей в традиционном нижнем регистре отображались

popovid replied:
хм, и правда, походу перепутал с  тегами db

popovid commented on internal/app/commands/logistic/location/command_delete.go:
```diff
@@ -0,0 +1,31 @@
+func (c *LogisticLocationCommander) Delete(inputMsg *tgbotapi.Message) {
+	args := inputMsg.CommandArguments()
+
+	id, err := strconv.Atoi(args)
```
тоже предложу сразу добавить функцию для парсинга uint

zaurbekzet replied:
Заменил: 3df50ad6c43b661e4b461f7eac4755ac771fb221

popovid commented on internal/service/logistic/location/service.go:
```diff
@@ -0,0 +1,127 @@
+		}
+	}
+
+	return nil, &LocationNotFoundError{locationID}
```
может воспользоваться errors.New?

zaurbekzet replied:
Передача несуществующего `locationID` встречается в коде неоднократно, поэтому решил создать кастомную ошибку и передавать в нее id

popovid commented on internal/service/logistic/location/service.go:
```diff
@@ -0,0 +1,127 @@
+	s.nextID++
+	s.locations = append(s.locations, location)
+
+	return location.ID, nil // TODO: Что делать с ошибкой?
```
а может быть добавить валидацию, что поля не пустые? в таком случае и ошибка появится)

zaurbekzet replied:
Добавил валидацию полей при создании и редактировании сущности: f45bd728a062ef67a03b83e6f50c588035a2c443

popovid commented on internal/service/logistic/location/service.go:
```diff
@@ -0,0 +1,127 @@
+	return fmt.Sprintf("location with ID %d not found", e.ID)
+}
+
+var initialLocations = []logistic.Location{
```
предложил бы переместить в файл location и переименовать его может в models?

zaurbekzet replied:
Согласен насчет перемещения `initialLocations` (b7011e6d995e8e5e54196168e40f6f8fc13b537d), но с переименованием файла _location.go_ в _models.go_ вопросы. Файл моделей _location.go_ располагается в папке _internal/model/logistic_ (согласно сниппетам из описания задания https://github.com/ozonmp/omp-docs), т.е. в папке домена без спецификации поддомена. Если переименовать его в _models.go_, то получится, что все модели поддоменов домена _logistic_ должны будут лежать в этом файле.

popovid replied:
согласен

popovid commented on internal/app/commands/logistic/location/commander.go:
```diff
@@ -0,0 +1,69 @@
+
+const logPrefix = "LogisticLocationCommander"
+
+type LocationCommander interface {
```
Сверился с заданием, надо будет поправить название LocationCommander на LogisticLocationCommander.
Как здесь https://github.com/ozonmp/omp-bot/blob/master/internal/app/commands/demo/subdomain/commander.go

zaurbekzet replied:
Просто в описании задания (https://github.com/ozonmp/omp-docs) указан шаблон `{Subdomain}Commander` без спецификации домена. К тому же `LocationCommander` — интерфейс, а `DemoSubdomainCommander` — просто структура (в этом поддомене не используется интерфейс). А вот структуру, реализующую интерфейс `LocationCommander`, я уже назвал `LogisticLocationCommander` по аналогии с `DemoSubdomainCommander`.

popovid replied:
Согласен, не посмотрел, что там структура, принято

zaurbekzet committed:
Format imports using goimports

zaurbekzet committed:
Add command arguments format to help message

zaurbekzet committed:
Change strconv.Atoi() to strconv.ParseUint() for uint64 values

zaurbekzet committed:
Fix listing locations

ErrEmptyList is returned on empty list and ErrEndOfList is returned when there are no more locations to list on the next page

zaurbekzet committed:
Add location fields validation

zaurbekzet committed:
Move initial locations to model file

popovid reviewed (approved):
Все оки
